File: src/app/src/main/java/com/huawei/training/java/activities/CourseDetailsActivity.java
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.huawei.training.java.activities;

import android.content.Intent;
import android.os.Bundle;

import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.huawei.agconnect.cloud.database.CloudDBZoneObject;
import com.huawei.agconnect.cloud.database.CloudDBZoneQuery;
import com.huawei.training.R;
import com.huawei.training.java.adapters.CourseDetailsListAdapter;
import com.huawei.training.java.database.CloudDbAction;
import com.huawei.training.java.database.CloudDbUiCallbackListener;
import com.huawei.training.java.database.tables.CourseDetailsTable;
import com.huawei.training.java.database.tables.CoursePlatformTable;
import com.huawei.training.java.database.tables.CoursesDocTable;
import com.huawei.training.java.database.tables.MyCoursesTable;
import com.huawei.training.java.database.tables.RecentlyViewedCoursesTable;
import com.huawei.training.databinding.ActivityCourseDetailsBinding;
import com.huawei.training.java.listeners.CourseTypeItemClick;
import com.huawei.training.java.models.CourseDetailsDataModel;
import com.huawei.training.java.utils.AppUtils;
import com.huawei.training.java.utils.LearningApplication;
import com.huawei.training.java.utils.video.Constants;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * @since 2020
 * @author Huawei DTSE India
 */
public class CourseDetailsActivity extends BaseActivity implements CloudDbUiCallbackListener {
    /**
     * The Binding.
     */
    ActivityCourseDetailsBinding binding;
    /**
     * The My list data.
     */
    ArrayList<CourseDetailsDataModel> myListData;
    /**
     * The Coursename.
     */
    private String coursename = "";
    /**
     * The Course id.
     */
    private int courseId = 0;
    /**
     * The Device platforms.
     */
    private String devicePlatforms = null;
    /**
     * The My course index.
     */
    private int myCourseIndex = 1;
    /**
     * The Recently viewed index.
     */
    private int recentlyViewedIndex = -1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_course_details);
        checkToFinishActivity(this.getIntent().getExtras());
        Intent intent = getIntent();
        coursename = intent.getStringExtra(Constants.COURSE_NAME);
        courseId = Integer.parseInt(Objects.requireNonNull(
                intent.getStringExtra(Constants.COURSE_ID)));
        ((LearningApplication) getApplication()).setRecentlyViewedCoresRefresh(false);
        setToolbar(coursename);
        init();
    }

    private void init() {

            try {
                if(cloudDbHelper!=null) {
                    cloudDbHelper.addCallBackListener(this);
                }
            } catch (NullPointerException e) {
        }
        binding.share.setOnClickListener(
                start -> {
                    appAnalytics.shareCourseClickEvent(coursename);
                    AppUtils.shareActivity(CourseDetailsActivity.this);
                });
        binding.addtolist.setOnClickListener(
                start -> {
                    if (!showLoginDialog()) {
                        if (userObj != null) {
                            MyCoursesTable myCourses = new MyCoursesTable();
                            myCourses.setUserId(userObj.getuId());
                            myCourses.setMyCourseId(myCourseIndex);
                            myCourses.setCourseId(courseId);
                            myCourses.setCourseName(coursename);
                            appAnalytics.addToMyCourseCourseClickEvent(
                                    userObj.getEmailId(), coursename);
                            cloudDbHelper.getCloudDbQueyCalls().upsertMyCourse(
                                    myCourses, CloudDbAction.GET_MY_COURSES);
                        }
                    }
                });

        AppUtils.showLoadingDialog(this);
        queryWithCourseId();
    }

    private void checkToFinishActivity(Bundle bundle) {
        if (bundle != null) {
            boolean isActivityToBeFinish = Objects.requireNonNull(this.getIntent().getExtras())
                    .getBoolean(Constants.FINISHSTATUS);
            if (isActivityToBeFinish) {
                finish();
            }
        }
    }

    // set course data
    private void setCourseData(List<CoursePlatformTable> coPlatformList) {
        myListData = new ArrayList<>();
        for (CoursePlatformTable coPlatform : coPlatformList) {
            if (devicePlatforms.contains(coPlatform.getPlatformId().toString())) {
                myListData.add(
                        new CourseDetailsDataModel(coPlatform.getPlatformName(),
                                coPlatform.getPlatformId()
                                        .toString(), Constants.PRICE, 3f));
            }
        }
    }

    /**
     * The Course item click.
     */
    CourseTypeItemClick courseItemClick = this::navigatetoCourseDetails;

    // navigate to course details page
    private void navigatetoCourseDetails(CourseDetailsDataModel model) {
        Intent intent = new Intent(CourseDetailsActivity.this,
                CourseContentDetailsActivity.class);
        intent.putExtra(Constants.COURSE_NAME, "" + model.getName());
        intent.putExtra(Constants.COURSE_ID, "" + courseId);
        startActivity(intent);
    }

    // get data from db
    private void queryWithCourseId() {
        CloudDBZoneQuery cloudDBZoneQuery = CloudDBZoneQuery.where(CourseDetailsTable.class);
        cloudDBZoneQuery.equalTo("courseId", courseId);
        cloudDbHelper.getCloudDbQueyCalls().queryCourseDetailsTable
                (cloudDBZoneQuery, CloudDbAction.GET_ALL_COURSES);
    }

    @Override
    public void onSuccessDbData(CloudDbAction cloudDbAction, List<CloudDBZoneObject> dataList) {
        switch (cloudDbAction) {
            case GET_COURSE_DOC:
                loadCourseDoc((CoursesDocTable) ((List) dataList).get(0));
                break;
            case GET_ALL_COURSES:
                loadAllCourses((List) dataList);
                break;
            case GET_COURSE_PLATFORMS:
                loadPlatformCourses((List) dataList);
                break;
            case GET_MY_COURSES:
                loadMyCourses((List) dataList);
                break;
            case GET_MY_COURSES_QUERY:
                loadMyCoursesQuery((List) dataList);
                queryRecentlyViewed(true);
                break;
            case GET_ALL_RECENTLY_VIEWED_COURSES:
                loadRecentCourses((List) dataList);
                break;
            case GET_RECENTLY_VIEWED_COURSES:
                getRecentlyViewedCourses(dataList);
                break;
            default:
                break;
        }
    }

    private void loadCourseDoc(CoursesDocTable coDoc) {
        binding.description.setText(coDoc.getDocDescription());
        devicePlatforms = coDoc.getDocPlatformsIds();
        cloudDbHelper.getCloudDbQueyCalls().queryCoursePlatformTable(null, CloudDbAction.GET_COURSE_PLATFORMS);
    }

    private void loadAllCourses(List allCoursesList) {
        CourseDetailsTable coDetails = (CourseDetailsTable) allCoursesList.get(0);
        CloudDBZoneQuery cloudDBZoneQuery = CloudDBZoneQuery.where(CoursesDocTable.class);
        cloudDBZoneQuery.equalTo("courseName", coDetails.getCourseDocId());
        cloudDbHelper.getCloudDbQueyCalls().queryCoursesDocTable(cloudDBZoneQuery, CloudDbAction.GET_COURSE_DOC);
    }

    private void getRecentlyViewedCourses(List<CloudDBZoneObject> dataList) {
        if (((List) dataList).isEmpty()) insertRecentlyViewedCourse(recentlyViewedIndex);
        AppUtils.cancelLoading();
    }

    private void loadPlatformCourses(List<CoursePlatformTable> coPlatformList) {
        setCourseData(coPlatformList);
        CourseDetailsListAdapter adapter = new CourseDetailsListAdapter(myListData, courseItemClick);
        binding.recyclerviewContentlist.setHasFixedSize(true);
        binding.recyclerviewContentlist.setLayoutManager(new LinearLayoutManager(this));
        binding.recyclerviewContentlist.setAdapter(adapter);
        queryMyCourses(-1);
    }

    private void loadMyCourses(List<MyCoursesTable> myCoursesList) {
        if (!myCoursesList.isEmpty()) {
            myCourseIndex = myCoursesList.get(0).getMyCourseId() + 1;
            queryMyCourses(courseId);
        } else {
            queryRecentlyViewed(true);
        }
    }

    private void loadMyCoursesQuery(List<MyCoursesTable> myCoursesQueryList) {
        if (!myCoursesQueryList.isEmpty()) {
            for (MyCoursesTable myCourses : myCoursesQueryList) {
                if (myCourses.getCourseId() == courseId) {
                    myCourseIndex = myCourses.getMyCourseId();
                    binding.addtolist.setEnabled(false);
                    break;
                }
            }
        }
    }

    private void loadRecentCourses(List<RecentlyViewedCoursesTable> viewedCourses) {
        if (!viewedCourses.isEmpty()) {
            recentlyViewedIndex = viewedCourses.get(0).getRecentlyViewedId() + 1;
            queryRecentlyViewed(false);
        } else {
            insertRecentlyViewedCourse(1);
        }
    }

    @Override
    public void onSuccessDbQueryMessage(CloudDbAction cloudDbAction, String message) {
        AppUtils.cancelLoading();
        if (message != null) {
            showToast(message);
        }
        if (CloudDbAction.GET_MY_COURSES == cloudDbAction) {
            binding.addtolist.setEnabled(false);
        }
    }

    @Override
    public void onFailureDbQueryMessage(CloudDbAction cloudDbAction, String message) {
        AppUtils.cancelLoading();
        if (message != null) {
            showToast(message);
        }
    }

    private void insertRecentlyViewedCourse(int index) {
        if (userObj != null) {
            if (userObj.getuId() != null) {
                RecentlyViewedCoursesTable recentlyViewedCourses = new RecentlyViewedCoursesTable();
                recentlyViewedCourses.setRecentlyViewedId(index);
                recentlyViewedCourses.setCourseId("" + courseId);
                recentlyViewedCourses.setCourseName(coursename);
                recentlyViewedCourses.setUserId(userObj.getuId());
                ((LearningApplication) getApplication()).setRecentlyViewedCoresRefresh(true);
                cloudDbHelper
                        .getCloudDbQueyCalls()
                        .upsertRecentlyViewedCourse(recentlyViewedCourses,
                                CloudDbAction.GET_RECENTLY_VIEWED_COURSES_MSG);
            }
        }
    }

    // get data from db
    private void queryMyCourses(int courseId) {
        if (userObj != null) {
            if (userObj.getuId() != null) {
                CloudDBZoneQuery cloudDBZoneQuery = CloudDBZoneQuery.where(MyCoursesTable.class);
                if (courseId > -1) {
                    cloudDBZoneQuery.orderByAsc("myCourseId");
                    cloudDBZoneQuery.contains("userId", userObj.getuId()).equalTo("courseId", courseId);
                    Objects.requireNonNull(this)
                            .cloudDbHelper
                            .getCloudDbQueyCalls()
                            .queryMyCoursesTable(cloudDBZoneQuery, CloudDbAction.GET_MY_COURSES_QUERY);
                } else {
                    cloudDBZoneQuery.orderByDesc("myCourseId");
                    Objects.requireNonNull(this)
                            .cloudDbHelper
                            .getCloudDbQueyCalls()
                            .queryMyCoursesTable(cloudDBZoneQuery, CloudDbAction.GET_MY_COURSES);
                }
            }
        } else {
            AppUtils.cancelLoading();
        }
    }

    //calling query for recently viewed courses
    private void queryRecentlyViewed(Boolean isAllList) {
        CloudDBZoneQuery cloudDBZoneQuery = CloudDBZoneQuery.where(RecentlyViewedCoursesTable.class);
        if (isAllList) {
            cloudDBZoneQuery.orderByDesc("recentlyViewedId");
            cloudDbHelper
                    .getCloudDbQueyCalls()
                    .queryRecentlyRecentlyViewedCoursesTable
                            (cloudDBZoneQuery, CloudDbAction.GET_ALL_RECENTLY_VIEWED_COURSES);
        } else {
            if (userObj.getuId() != null) {
                cloudDBZoneQuery.equalTo("userId", userObj.getuId()).equalTo("courseId", "" + courseId);
                cloudDbHelper
                        .getCloudDbQueyCalls()
                        .queryRecentlyRecentlyViewedCoursesTable
                                (cloudDBZoneQuery, CloudDbAction.GET_RECENTLY_VIEWED_COURSES);
            }
        }
    }
}

File: src/app/src/main/java/com/huawei/training/kotlin/activities/CourseDetailsActivity.kt
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package com.huawei.training.kotlin.activities

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import com.huawei.agconnect.cloud.database.CloudDBZoneObject
import com.huawei.agconnect.cloud.database.CloudDBZoneQuery
import com.huawei.training.R
import com.huawei.training.kotlin.database.CloudDbHelper
import com.huawei.training.kotlin.adapters.CourseDetailsListAdapter
import com.huawei.training.kotlin.database.CloudDbAction
import com.huawei.training.kotlin.database.CloudDbUiCallbackListener
import com.huawei.training.kotlin.database.tables.*
import com.huawei.training.kotlin.listeners.CourseTypeItemClick
import com.huawei.training.kotlin.models.CourseDetailsDataModel
import com.huawei.training.kotlin.utils.AppUtils
import com.huawei.training.kotlin.utils.LearningApplication
import com.huawei.training.kotlin.utils.video.Constants
import kotlinx.android.synthetic.main.activity_course_details.*
import java.util.*

/**
 * @since 2020
 * @author Huawei DTSE India
 */
class CourseDetailsActivity : BaseActivity(), CloudDbUiCallbackListener {

    /**
     * The My list data.
     */
    var myListData: ArrayList<CourseDetailsDataModel>? = null

    /**
     * The Coursename.
     */
    private var coursename: String? = ""

    /**
     * The Course id.
     */
    private var courseId = 0

    /**
     * The Device platforms.
     */
    private var devicePlatforms: String? = null

    /**
     * The My course index.
     */
    private var myCourseIndex = 1

    /**
     * The Recently viewed index.
     */
    private var recentlyViewedIndex = -1

    /**
     * Gets cloud db helper.
     *
     * @return the cloud db helper
     */
    var cloudDbHelper: CloudDbHelper?=null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_course_details)
        checkToFinishActivity(this.intent.extras)
        val intent = intent
        coursename = intent.getStringExtra(Constants.COURSE_NAME)
        courseId = intent.getStringExtra(Constants.COURSE_ID)?.toInt()!!
        (application as LearningApplication).isRecentlyViewedCoresRefresh = false
        setToolbar(coursename)
        init()
    }

    private fun init() {
        try {
            cloudDbHelper= CloudDbHelper.getInstance(applicationContext)
            if (cloudDbHelper != null) {
                cloudDbHelper?.addCallBackListener(this)
            }
        } catch (e: NullPointerException) {
        }
        share.setOnClickListener { start: View? ->
            appAnalytics?.shareCourseClickEvent(coursename)
            AppUtils.shareActivity(this@CourseDetailsActivity)
        }
        addtolist.setOnClickListener { start: View? ->
            if (!showLoginDialog()) {
                if (userObj != null) {
                    val myCourses = MyCoursesTable()
                    myCourses.userId = userObj!!.getuId()
                    myCourses.myCourseId = myCourseIndex
                    myCourses.courseId = courseId
                    myCourses.courseName = coursename
                    appAnalytics?.addToMyCourseCourseClickEvent(
                            userObj?.emailId, coursename)
                    cloudDbHelper?.getCloudDbAllQueyCalls()?.upsertMyCourse(
                            myCourses, CloudDbAction.GET_MY_COURSES)
                }
            }
        }
        AppUtils.showLoadingDialog(this)
        queryWithCourseId()
    }

    private fun checkToFinishActivity(bundle: Bundle?) {
        if (bundle != null) {
            val isActivityToBeFinish = Objects.requireNonNull(this.intent.extras)
                    ?.getBoolean(Constants.FINISHSTATUS)
            if (isActivityToBeFinish!!) {
                finish()
            }
        }
    }

    // set course data
    private fun setCourseData(coPlatformList: List<CoursePlatformTable>) {
        myListData = ArrayList()
        for (coPlatform in coPlatformList) {
            if (devicePlatforms!!.contains(coPlatform.platformId.toString())) {
                myListData?.add(
                        CourseDetailsDataModel(coPlatform.platformName.toString(),
                                coPlatform.platformId
                                        .toString(), Constants.PRICE, 3f))
            }
        }
    }

    /**
     * The Course item click.
     */
    var courseItemClick = object : CourseTypeItemClick {
        override fun courseOnClick(model: CourseDetailsDataModel?) {
            if (model != null) {
                navigatetoCourseDetails(model)
            }
        }
    }

    // navigate to course details page
    private fun navigatetoCourseDetails(model: CourseDetailsDataModel) {
        val intent = Intent(this@CourseDetailsActivity,
                CourseContentDetailsActivity::class.java)
        intent.putExtra(Constants.COURSE_NAME, "" + model.name)
        intent.putExtra(Constants.COURSE_ID, "" + courseId)
        startActivity(intent)
    }

    // get data from db
    private fun queryWithCourseId() {
        val cloudDBZoneQuery: CloudDBZoneQuery<*> = CloudDBZoneQuery.where(CourseDetailsTable::class.java)
        cloudDBZoneQuery.equalTo("courseId", courseId)
        cloudDbHelper?.getCloudDbAllQueyCalls()?.queryCourseDetailsTable(cloudDBZoneQuery as CloudDBZoneQuery<CourseDetailsTable>?, CloudDbAction.GET_ALL_COURSES)
    }

    override fun onSuccessDbData(cloudDbAction: CloudDbAction?, dataList: List<CloudDBZoneObject?>?) {
        when (cloudDbAction) {
            CloudDbAction.GET_COURSE_DOC -> loadCourseDoc(dataList?.get(0) as CoursesDocTable)
            CloudDbAction.GET_ALL_COURSES -> loadAllCourses(dataList as List<*>)
            CloudDbAction.GET_COURSE_PLATFORMS -> loadPlatformCourses(dataList as List<CoursePlatformTable>)
            CloudDbAction.GET_MY_COURSES -> loadMyCourses(dataList as List<MyCoursesTable>)
            CloudDbAction.GET_MY_COURSES_QUERY -> {
                loadMyCoursesQuery(dataList as List<MyCoursesTable>)
                queryRecentlyViewed(true)
            }
            CloudDbAction.GET_ALL_RECENTLY_VIEWED_COURSES -> loadRecentCourses(dataList as List<RecentlyViewedCoursesTable>)
            CloudDbAction.GET_RECENTLY_VIEWED_COURSES -> getRecentlyViewedCourses(dataList as List<CloudDBZoneObject>)
            else -> {
            }
        }
    }

    private fun loadCourseDoc(coDoc: CoursesDocTable) {
        description.text = coDoc.docDescription
        devicePlatforms = coDoc.docPlatformsIds
        cloudDbHelper?.getCloudDbAllQueyCalls()?.queryCoursePlatformTable(null, CloudDbAction.GET_COURSE_PLATFORMS)
    }

    private fun loadAllCourses(allCoursesList: List<*>) {
        val coDetails = allCoursesList[0] as CourseDetailsTable
        val cloudDBZoneQuery: CloudDBZoneQuery<*> = CloudDBZoneQuery.where(CoursesDocTable::class.java)
        cloudDBZoneQuery.equalTo("courseName", coDetails.courseDocId)
        cloudDbHelper?.getCloudDbAllQueyCalls()?.queryCoursesDocTable(cloudDBZoneQuery as CloudDBZoneQuery<CoursesDocTable>?, CloudDbAction.GET_COURSE_DOC)
    }

    private fun getRecentlyViewedCourses(dataList: List<CloudDBZoneObject>) {
        if ((dataList as List<*>).isEmpty()) insertRecentlyViewedCourse(recentlyViewedIndex)
        AppUtils.cancelLoading()
    }

    private fun loadPlatformCourses(coPlatformList: List<CoursePlatformTable>) {
        setCourseData(coPlatformList)
        val adapter = myListData?.let { CourseDetailsListAdapter(it, courseItemClick) }
        recyclerview_contentlist.setHasFixedSize(true)
        recyclerview_contentlist.layoutManager = LinearLayoutManager(this)
        recyclerview_contentlist.adapter = adapter
        queryMyCourses(-1)
    }

    private fun loadMyCourses(myCoursesList: List<MyCoursesTable>) {
        if (!myCoursesList.isEmpty()) {
            myCourseIndex = myCoursesList[0].myCourseId!! + 1
            queryMyCourses(courseId)
        } else {
            queryRecentlyViewed(true)
        }
    }

    private fun loadMyCoursesQuery(myCoursesQueryList: List<MyCoursesTable>) {
        if (!myCoursesQueryList.isEmpty()) {
            for (myCourses in myCoursesQueryList) {
                if (myCourses.courseId == courseId) {
                    myCourseIndex = myCourses.myCourseId!!
                    addtolist.isEnabled = false
                    break
                }
            }
        }
    }

    private fun loadRecentCourses(viewedCourses: List<RecentlyViewedCoursesTable>) {
        if (!viewedCourses.isEmpty()) {
            recentlyViewedIndex = (viewedCourses[0].recentlyViewedId?.plus(1)!!)
            queryRecentlyViewed(false)
        } else {
            insertRecentlyViewedCourse(1)
        }
    }

    override fun onSuccessDbQueryMessage(cloudDbAction: CloudDbAction?, message: String?) {
        AppUtils.cancelLoading()
        message?.let { showToast(it) }
        if (CloudDbAction.GET_MY_COURSES == cloudDbAction) {
            addtolist.isEnabled = false
        }
    }

    override fun onFailureDbQueryMessage(cloudDbAction: CloudDbAction?, message: String?) {
        AppUtils.cancelLoading()
        message?.let { showToast(it) }
    }

    private fun insertRecentlyViewedCourse(index: Int) {
        if (userObj != null) {
            if (userObj?.getuId() != null) {
                val recentlyViewedCourses = RecentlyViewedCoursesTable()
                recentlyViewedCourses.recentlyViewedId = index
                recentlyViewedCourses.courseId = "" + courseId
                recentlyViewedCourses.courseName = coursename
                recentlyViewedCourses.userId = userObj!!.getuId()
                (application as LearningApplication).isRecentlyViewedCoresRefresh = true
                cloudDbHelper
                        ?.getCloudDbAllQueyCalls()
                        ?.upsertRecentlyViewedCourse(recentlyViewedCourses,
                                CloudDbAction.GET_RECENTLY_VIEWED_COURSES_MSG)
            }
        }
    }

    // get data from db
    private fun queryMyCourses(courseId: Int) {
        if (userObj != null) {
            if (userObj?.getuId() != null) {
                val cloudDBZoneQuery: CloudDBZoneQuery<*> = CloudDBZoneQuery.where(MyCoursesTable::class.java)
                if (courseId > -1) {
                    cloudDBZoneQuery.orderByAsc("myCourseId")
                    userObj?.getuId()?.let { cloudDBZoneQuery.contains("userId", it).equalTo("courseId", courseId) }
                    Objects.requireNonNull(this).cloudDbHelper
                            ?.getCloudDbAllQueyCalls()
                            ?.queryMyCoursesTable(cloudDBZoneQuery as CloudDBZoneQuery<MyCoursesTable>?, CloudDbAction.GET_MY_COURSES_QUERY)
                } else {
                    cloudDBZoneQuery.orderByDesc("myCourseId")
                    Objects.requireNonNull(this).cloudDbHelper
                            ?.getCloudDbAllQueyCalls()
                            ?.queryMyCoursesTable(cloudDBZoneQuery as CloudDBZoneQuery<MyCoursesTable>?, CloudDbAction.GET_MY_COURSES)
                }
            }
        } else {
            AppUtils.cancelLoading()
        }
    }

    //calling query for recently viewed courses
    private fun queryRecentlyViewed(isAllList: Boolean) {
        val cloudDBZoneQuery: CloudDBZoneQuery<*> = CloudDBZoneQuery.where(RecentlyViewedCoursesTable::class.java)
        if (isAllList) {
            cloudDBZoneQuery.orderByDesc("recentlyViewedId")
            cloudDbHelper
                    ?.getCloudDbAllQueyCalls()
                    ?.queryRecentlyRecentlyViewedCoursesTable(cloudDBZoneQuery as CloudDBZoneQuery<RecentlyViewedCoursesTable>?, CloudDbAction.GET_ALL_RECENTLY_VIEWED_COURSES)
        } else {
            if (userObj?.getuId() != null) {
                cloudDBZoneQuery.equalTo("userId", userObj?.getuId()).equalTo("courseId", "" + courseId)
                cloudDbHelper?.getCloudDbAllQueyCalls()
                        ?.queryRecentlyRecentlyViewedCoursesTable(cloudDBZoneQuery as CloudDBZoneQuery<RecentlyViewedCoursesTable>?, CloudDbAction.GET_RECENTLY_VIEWED_COURSES)
            }
        }
    }
}